File: nodecore-spv/src/main/java/veriblock/model/TransactionMeta.java
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package veriblock.model;

import org.veriblock.sdk.models.Sha256Hash;
import org.veriblock.sdk.models.VBlakeHash;
import veriblock.listeners.TransactionStateChangedListener;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executor;

public class TransactionMeta {
    private final List<VBlakeHash> appearsInBlock;
    private int appearsAtChainHeight = -1;
    private int depth;
    private final Set<String> seenByPeers;
    private final CopyOnWriteArrayList<ListenerRegistration<TransactionStateChangedListener>> stateChangedListeners;
    private final Sha256Hash txId;
    private MetaState state = MetaState.UNKNOWN;
    private VBlakeHash appearsInBestChainBlock;

    public TransactionMeta(Sha256Hash txId) {
        this.txId = txId;
        this.appearsInBlock = new ArrayList<>();
        this.stateChangedListeners = new CopyOnWriteArrayList<>();
        this.seenByPeers = new HashSet<>();
    }

    public VBlakeHash getAppearsInBestChainBlock() {
        return appearsInBestChainBlock;
    }

    public void setAppearsInBestChainBlock(VBlakeHash appearsInBestChainBlock) {
        this.appearsInBestChainBlock = appearsInBestChainBlock;
    }


    public List<VBlakeHash> getAppearsInBlock() {
        return appearsInBlock;
    }

    public void addBlockAppearance(VBlakeHash hash) {
        appearsInBlock.add(hash);
    }


    public int getAppearsAtChainHeight() {
        return appearsAtChainHeight;
    }

    public void setAppearsAtChainHeight(int appearsAtChainHeight) {
        this.appearsAtChainHeight = appearsAtChainHeight;
    }

    public int getDepth() {
        return depth;
    }

    public void setDepth(int depth) {
        this.depth = depth;
    }

    public void incrementDepth() {
        this.depth++;
        informListenersDepthChanged();
    }

    public int getBroadcastPeerCount() {
        return seenByPeers.size();
    }

    public boolean recordBroadcast(String peer) {
        boolean added = seenByPeers.add(peer);
        if (!added) {
            return false;
        }

        if (getState() == MetaState.UNKNOWN) {
            setState(MetaState.PENDING);
        }

        return true;
    }

    public void addTransactionStateChangedListener(TransactionStateChangedListener listener, Executor executor) {
        stateChangedListeners.add(new ListenerRegistration<>(listener, executor));
    }

    public void removeTransactionStateChangedListener(TransactionStateChangedListener listener) {
        ListenerRegistration.removeFromList(listener, stateChangedListeners);
    }

    private void informListenersStateChanged() {
        TransactionMeta self = this;
        for (final ListenerRegistration<TransactionStateChangedListener> registration : stateChangedListeners) {
            registration.executor.execute(() -> registration.listener.onTransactionStateChanged(self,
                    TransactionStateChangedListener.ChangeReason.STATE));
        }
    }

    private void informListenersDepthChanged() {
        TransactionMeta self = this;
        for (final ListenerRegistration<TransactionStateChangedListener> registration : stateChangedListeners) {
            registration.executor.execute(() -> registration.listener.onTransactionStateChanged(self,
                    TransactionStateChangedListener.ChangeReason.DEPTH));
        }
    }

    public Sha256Hash getTxId() {
        return txId;
    }

    public MetaState getState() {
        return state;
    }

    public void setState(MetaState state) {
        if (this.state == state) return;

        this.state = state;
        if (state == MetaState.PENDING) {
            depth = 0;
            appearsAtChainHeight = -1;
        }
        informListenersStateChanged();
    }

    public enum MetaState {
        PENDING(1),
        CONFIRMED(2),
        DEAD(3),
        UNKNOWN(0);

        private int value;

        MetaState(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

        public static MetaState forNumber(int value) {
            switch (value) {
                case 0:
                    return UNKNOWN;
                case 1:
                    return PENDING;
                case 2:
                    return CONFIRMED;
                case 3:
                    return DEAD;
                default:
                    return null;
            }
        }
    }

}

File: nodecore-spv/src/main/java/veriblock/model/TransactionMeta.kt
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package veriblock.model

import org.veriblock.sdk.models.Sha256Hash
import org.veriblock.sdk.models.VBlakeHash
import veriblock.listeners.TransactionStateChangedListener
import java.util.ArrayList
import java.util.HashSet
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.Executor

class TransactionMeta(
    val txId: Sha256Hash
) {
    private val appearsInBlock: MutableList<VBlakeHash> = ArrayList()
    var appearsAtChainHeight = -1
    var depth = 0
    private val seenByPeers: MutableSet<String> = HashSet()
    private val stateChangedListeners: CopyOnWriteArrayList<ListenerRegistration<TransactionStateChangedListener>> = CopyOnWriteArrayList()
    var state = MetaState.UNKNOWN
        private set
    var appearsInBestChainBlock: VBlakeHash? = null

    fun getAppearsInBlock(): List<VBlakeHash> {
        return appearsInBlock
    }

    fun addBlockAppearance(hash: VBlakeHash) {
        appearsInBlock.add(hash)
    }

    fun incrementDepth() {
        depth++
        informListenersDepthChanged()
    }

    val broadcastPeerCount: Int
        get() = seenByPeers.size

    fun recordBroadcast(peer: String): Boolean {
        val added = seenByPeers.add(peer)
        if (!added) {
            return false
        }
        if (state == MetaState.UNKNOWN) {
            state = MetaState.PENDING
        }
        return true
    }

    fun addTransactionStateChangedListener(listener: TransactionStateChangedListener, executor: Executor) {
        stateChangedListeners.add(ListenerRegistration(listener, executor))
    }

    fun removeTransactionStateChangedListener(listener: TransactionStateChangedListener) {
        ListenerRegistration.Companion.removeFromList<TransactionStateChangedListener>(listener, stateChangedListeners)
    }

    private fun informListenersStateChanged() {
        val self = this
        for (registration in stateChangedListeners) {
            registration.executor.execute {
                registration.listener.onTransactionStateChanged(
                    self,
                    TransactionStateChangedListener.ChangeReason.STATE
                )
            }
        }
    }

    private fun informListenersDepthChanged() {
        val self = this
        for (registration in stateChangedListeners) {
            registration.executor.execute {
                registration.listener.onTransactionStateChanged(
                    self,
                    TransactionStateChangedListener.ChangeReason.DEPTH
                )
            }
        }
    }

    fun setState(state: MetaState) {
        if (this.state == state) {
            return
        }
        this.state = state
        if (state == MetaState.PENDING) {
            depth = 0
            appearsAtChainHeight = -1
        }
        informListenersStateChanged()
    }

    enum class MetaState(
        val value: Int
    ) {
        PENDING(1),
        CONFIRMED(2),
        DEAD(3),
        UNKNOWN(0);

        companion object {
            fun forNumber(value: Int): MetaState {
                return when (value) {
                    0 -> UNKNOWN
                    1 -> PENDING
                    2 -> CONFIRMED
                    3 -> DEAD
                    else -> throw IllegalArgumentException("There's no MetaState for value $value")
                }
            }
        }

    }
}
